File: packages/python/slop-ai/src/slop_ai/store.py
"""Helpers for binding application stores to SLOP registrations."""

from __future__ import annotations

from threading import RLock, Timer
from typing import Any, Callable, Protocol, TypeVar, runtime_checkable

StateT = TypeVar("StateT")
StoreUnsubscribe = Callable[[], None]
StorePath = str | Callable[[StateT], str]


@runtime_checkable
class StoreSubscription(Protocol):
    """Subscription object returned by some state libraries."""

    def unsubscribe(self) -> None: ...


StoreSubscribeResult = StoreUnsubscribe | StoreSubscription


@runtime_checkable
class StateStore(Protocol[StateT]):
    """Minimal store shape supported by :func:`expose_store`."""

    def get_state(self) -> StateT: ...
    def subscribe(self, listener: Callable[[], None]) -> StoreSubscribeResult: ...


@runtime_checkable
class StoreTarget(Protocol):
    """Minimal SLOP registration target supported by :func:`expose_store`."""

    def register(self, path: str, descriptor: dict[str, Any]) -> None: ...
    def unregister(self, path: str, *, recursive: bool = False) -> None: ...


def expose_store(
    target: StoreTarget,
    path: StorePath[StateT],
    store: StateStore[StateT],
    project: Callable[[StateT], dict[str, Any]],
    *,
    equals: Callable[[StateT, StateT], bool] | None = None,
    debounce_ms: int = 0,
) -> StoreUnsubscribe:
    """Bind a generic state store to a SLOP node.

    The store supplies change notifications; ``project`` decides what semantic
    state and affordances to expose. This works with any Python store object
    that has ``get_state()`` plus ``subscribe(listener)``.
    """

    lock = RLock()
    current_path: str | None = None
    previous_state: StateT | None = None
    has_previous_state = False
    disposed = False
    timer: Timer | None = None

    def update() -> None:
        nonlocal current_path, previous_state, has_previous_state

        with lock:
            if disposed:
                return

            state = store.get_state()
            if has_previous_state and equals is not None and equals(previous_state, state):  # type: ignore[arg-type]
                return

            next_path = path(state) if callable(path) else path
            if current_path is not None and current_path != next_path:
                target.unregister(current_path, recursive=True)

            target.register(next_path, project(state))
            current_path = next_path
            previous_state = state
            has_previous_state = True

    def schedule_update() -> None:
        nonlocal timer

        with lock:
            if disposed:
                return
            if debounce_ms <= 0:
                update()
                return
            if timer is not None:
                timer.cancel()
            timer = Timer(debounce_ms / 1000, update)
            timer.daemon = True
            timer.start()

    update()
    subscription = store.subscribe(schedule_update)

    def dispose() -> None:
        nonlocal current_path, disposed, timer

        with lock:
            disposed = True
            if timer is not None:
                timer.cancel()
                timer = None
            _dispose_subscription(subscription)
            if current_path is not None:
                target.unregister(current_path, recursive=True)
                current_path = None

    return dispose


def _dispose_subscription(subscription: StoreSubscribeResult) -> None:
    if callable(subscription):
        subscription()
    else:
        subscription.unsubscribe()

File: packages/typescript/sdk/core/src/store.ts
import type { NodeDescriptor } from "./types";

export type StoreUnsubscribe = () => void;

export interface StoreSubscription {
  unsubscribe(): void;
}

export type StoreSubscribeResult = StoreUnsubscribe | StoreSubscription;

export interface StateStore<S> {
  getState(): S;
  subscribe(listener: () => void): StoreSubscribeResult;
}

export interface StoreTarget {
  register(path: string, descriptor: NodeDescriptor): void;
  unregister(path: string, opts?: { recursive?: boolean }): void;
}

export type StorePath<S> = string | ((state: S) => string);

export interface ExposeStoreOptions<S> {
  /**
   * Skip re-registering when the store emits but the selected state is
   * equivalent for this SLOP projection.
   */
  equals?: (previous: S, next: S) => boolean;
  /**
   * Debounce store emissions before recomputing the descriptor. The SLOP
   * client still batches tree rebuilds; this controls projection work.
   */
  debounceMs?: number;
}

/**
 * Bind a generic state store to a SLOP node.
 *
 * The store supplies change notifications; the projection decides what
 * semantic state and affordances to expose. This works with Zustand, Redux,
 * and any other store that has `getState()` plus `subscribe()`.
 */
export function exposeStore<S>(
  target: StoreTarget,
  path: StorePath<S>,
  store: StateStore<S>,
  project: (state: S) => NodeDescriptor,
  options: ExposeStoreOptions<S> = {},
): StoreUnsubscribe {
  let currentPath: string | undefined;
  let previousState: S | undefined;
  let hasPreviousState = false;
  let disposed = false;
  let timer: ReturnType<typeof setTimeout> | undefined;

  const update = () => {
    if (disposed) return;

    const state = store.getState();
    if (hasPreviousState && options.equals?.(previousState as S, state)) {
      return;
    }

    const nextPath = resolveStorePath(path, state);
    if (currentPath && currentPath !== nextPath) {
      target.unregister(currentPath, { recursive: true });
    }

    target.register(nextPath, project(state));
    currentPath = nextPath;
    previousState = state;
    hasPreviousState = true;
  };

  const scheduleUpdate = () => {
    if (disposed) return;

    const debounceMs = options.debounceMs ?? 0;
    if (debounceMs <= 0) {
      update();
      return;
    }

    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
      timer = undefined;
      update();
    }, debounceMs);
  };

  update();

  const subscription = store.subscribe(scheduleUpdate);

  return () => {
    disposed = true;
    if (timer) {
      clearTimeout(timer);
      timer = undefined;
    }
    disposeStoreSubscription(subscription);
    if (currentPath) {
      target.unregister(currentPath, { recursive: true });
      currentPath = undefined;
    }
  };
}

function resolveStorePath<S>(path: StorePath<S>, state: S): string {
  return typeof path === "function" ? path(state) : path;
}

function disposeStoreSubscription(subscription: StoreSubscribeResult): void {
  if (typeof subscription === "function") {
    subscription();
    return;
  }
  subscription.unsubscribe();
}
